src/libasr/codegen/llvm_utils.cpp

```diff
@@ -250,15 +304,73 @@ namespace LFortran {
         copy_data = builder->CreateBitCast(copy_data, el_type->getPointerTo());
-        builder->CreateMemCpy(copy_data, llvm::MaybeAlign(), src_data,
-                              llvm::MaybeAlign(), arg_size);
-        builder->CreateStore(copy_data, get_pointer_to_list_data(dest));
+
+        // We consider the case when the element type of a list is defined by a struct
+        // which may also contain non-trivial structs (such as in case of list[list[f64]],
+        // list[tuple[f64]]). We need to make sure that all the data inside those structs
+        // is deepcopied and not just the address of the first element of those structs.
+        // Hence we dive deeper into the lowest level of nested types and deepcopy everything
+        // properly. If we don't consider this case then the data only from first level of nested types
+        // will be deep copied and rest will be shallow copied. The importance of this case
+        // can be figured out by goind through, integration_tests/test_list_06.py and
+        // integration_tests/test_list_07.py.
+        if( LLVM::is_llvm_struct(list_type->m_type) ) {
+            builder->CreateStore(copy_data, get_pointer_to_list_data(dest));
+            llvm::AllocaInst *pos_ptr = builder->CreateAlloca(llvm::Type::getInt32Ty(context),
+                                                              nullptr);
+            LLVM::CreateStore(*builder, llvm::ConstantInt::get(llvm::Type::getInt32Ty(context),
+                                                               llvm::APInt(32, 0)), pos_ptr);
+
+            llvm::BasicBlock *loophead = llvm::BasicBlock::Create(context, "loop.head");
+            llvm::BasicBlock *loopbody = llvm::BasicBlock::Create(context, "loop.body");
+            llvm::BasicBlock *loopend = llvm::BasicBlock::Create(context, "loop.end");
+
+            // head
+            llvm_utils->start_new_block(loophead);
+            {
+                llvm::Value *cond = builder->CreateICmpSGT(
+                                            src_end_point,
+                                            LLVM::CreateLoad(*builder, pos_ptr));
+                builder->CreateCondBr(cond, loopbody, loopend);
+            }
+
+            // body
+            llvm_utils->start_new_block(loopbody);
+            {
+                llvm::Value* pos = LLVM::CreateLoad(*builder, pos_ptr);
+                llvm::Value* srci = read_item(src, pos, true);
+                llvm::Value* desti = read_item(dest, pos, true);
+                llvm_utils->deepcopy(srci, desti, list_type->m_type, module);
+                llvm::Value* tmp = builder->CreateAdd(
+                            pos,
+                            llvm::ConstantInt::get(context, llvm::APInt(32, 1)));
+                LLVM::CreateStore(*builder, tmp, pos_ptr);
+            }
+
+            builder->CreateBr(loophead);
+
+            // end
+            llvm_utils->start_new_block(loopend);
```

Comment:
Does this slow anything down in the already existing list (or any other) functionality in master? For example in our current benchmarks for lists? Or is this only used for new features implemented in this PR, but does not affect any feature already in master.

Reply:
This code path will only be taken for nesting of types inside lists. For `list[f64]` (i.e., a list with intrinsic element type, the code path in `main` will be taken). So, theoretically no change in benchmarks should happen.

Reply:
When you say "code path", you mean code path at compile time? If so, then this is good.

If you mean "runtime", then that must mean there might be some "if statement" being executed at runtime, and we don't want that.

Reply:
> When you say "code path", you mean code path at compile time? 

Yes code path at compile time.

Reply:
Updated the benchmarks for my linux system in https://github.com/lcompilers/lpython/issues/857#issuecomment-1200960150. I don't see any change with `list05` (the branch of this PR).

Reply:
Updated https://github.com/lcompilers/lpython/issues/857#issuecomment-1200960150 with results from my macbook, this and main branch give no difference for me.

Reply:
Bottomline - This and main branch give same results on both of my systems. So, theoretically and practically from my perspective this PR wouldn't slow down anything for features already existing in `main`. @certik Do you want to merge it?